casper/src/main/resources/RevVault.rho

```diff
@@ -175,100 +219,139 @@ in {
-                  for (@result <- ret2) {
-                    ret!(result) |
-                    for (logCh <<- logStore) {
-                      if (Nil != *logCh) {
-                        new bd(`rho:block:data`), bdCh in {
-                          bd!(*bdCh) |
-                          for (@blockNumber, @timestamp, @sender <- bdCh) {
-                            logCh!(["transfer", targetAddress, amount, result, blockNumber, timestamp, sender])
+          // Supplies balance, transfer, and readEntry methods on `revVault`.
+          contract _newVault(revVault, @fromRevAddress, purse) = {
+            // Returns `revVault` balance on ret
+            contract revVault(@"balance", ret) = {
+              purse!("getBalance", *ret)
+            } |
+            // Transfers `amount` from `revVault` (validated with `authKey`) to the vault at `toRevAddress`
+            // Returns either (true, Nil) or (false, "BUG FOUND: purse deposit failed") on `ret`
+            // Logs successful transfers on a read-ony channel made available through "getLog" method
+            contract revVault(@"transfer", @toRevAddress, @amount, authKey, ret) = {
+              new ret2 in {
+                _transferTemplate!(fromRevAddress, *purse, toRevAddress, amount, *authKey, *ret2) |
+                for (@result <- ret2) {
+                  if (result.nth(0)) {
+                    new
+                      bd(`rho:block:data`),
+                      bdCh
+                    in {
+                      bd!(*bdCh) |
+                      for (@blockNumber, @timestamp, @sender <- bdCh) {
+                        // Log the transfer in both vault logs with the same unique id
+                        new id in {
+                          // Record new log entry in both vault logs
+                          match ("transfer",
+                            { "id"     : *id,
+                              "from"   : fromRevAddress,
+                              "to"     : toRevAddress,
+                              "amount" : amount,
+                              "block"  : { "number" : blockNumber, "timestamp" : timestamp, "sender" : sender }
+                            }
+                          ) {
+                            entryTuple => {
+                              new
+                                fromLogRet,
+                                toLogRet
+                              in {
+                                // Lookup vault associated with `toRevAddress`
+                                TreeHashMap!("get", vaultMap, toRevAddress, *toLogRet) |
+                                // Lookup log channel associated with `fromRevAddress`
+                                TreeHashMap!("get", vaultMap, fromRevAddress, *fromLogRet) |
+                                for (@(_, _, *fromLogCh) <- fromLogRet;
+                                     @(_, _, *toLogCh)   <- toLogRet) {
```

Comment:
It is possible the `toAddress` vault is not created and the `get` above would return `Nil`. 
In this case the transfer is done and the `ret` won't return anything and the log would be recorded.